### ISSUE TITLE
CUDA build broken on master

### ISSUE THREAD

**Issue by kpeace** [mesonbuild/meson#5870]
As of commit 3fbb45016610deba863630b20e95f60d2d49a487 CUDA compilation is broken.

Sample code (can be cloned from: https://github.com/kpeace/meson-cuda.git)
hello.cu:
``` cpp
#include <iostream>
#include <cuda_runtime.h>


__global__ 
void hello(int *b) 
{
    *b = 43;
}


int main(){


    int *pd;
    cudaMalloc(&pd, sizeof(int));
    //pd = (int*)malloc(sizeof(int));
    hello<<<1, 1>>>(pd);
    cudaDeviceSynchronize();

    int res = 0;
    cudaMemcpy(&res, pd, sizeof(int), cudaMemcpyDeviceToHost );
    cudaFree(pd);
    //res = *pd;

    std::cout << "hello world: " << res << std::endl;
    return 0;
}
```

meson.build
``` meson
project('hello-cuda', 'cpp', 'cuda')

cuda_comp = meson.get_compiler('cuda')
cuda = import('unstable-cuda')
nvcc_flags = cuda.nvcc_arch_flags(cuda_comp, 'Auto')
executable('hello-cuda', 'hello.cu', cuda_args : nvcc_flags)
```

Using the stable 0.51.2 this project builds and compiles properly.
Using the master from git `meson build src` failes with:
`meson.build:1:0: ERROR: Unable to determine dynamic linker.`

Full meson-log.txt:
```
Build started at 2019-09-01T15:22:10.420621
Main binary: /usr/bin/python3
Python system: Linux
The Meson build system
Version: 0.51.999
Source dir: /home/eyeway/cuda-workspace/meson-cuda/src
Build dir: /home/eyeway/cuda-workspace/meson-cuda/build
Build type: native build
Project name: hello-cuda
Project version: undefined
No CXXFLAGS in the environment, not changing global flags.
No LDFLAGS in the environment, not changing global flags.
No CPPFLAGS in the environment, not changing global flags.
Sanity testing C++ compiler: c++
Is cross compiler: False.
Sanity check compiler command line: c++ /home/eyeway/cuda-workspace/meson-cuda/build/meson-private/sanitycheckcpp.cc -o /home/eyeway/cuda-workspace/meson-cuda/build/meson-private/sanitycheckcpp.exe -pipe -D_FILE_OFFSET_BITS=64
Sanity check compile stdout:

-----
Sanity check compile stderr:

-----
Running test binary command: /home/eyeway/cuda-workspace/meson-cuda/build/meson-private/sanitycheckcpp.exe
C++ compiler for the build machine: c++ (gcc 7.4.0 "c++ (Ubuntu/Linaro 7.4.0-1ubuntu1~18.04.1) 7.4.0")
C++ linker for the build machine: GNU ld.bfd 2.30

meson.build:1:0: ERROR: Unable to determine dynamic linker.
```
I done a quick `git bisect` to find the offending commit and it seams it stopped working at 3fbb45016610deba863630b20e95f60d2d49a487.

Looking at the commit, it seams that no linker was defined for CUDA (should be nvcc). But I am not familiar enough with the code to understand exactly what change is needed to fix this.

If any one can point out what is needed I will be happy to make the change and test it.

**Comment by kpeace** [comment]
The issue might not be with identifying the linker. The issue might be with identifying Cuda and choosing nvcc as the compiler.

I went back and checked the working 0.51.2 release and looked at the meson-log.txt:
```
Build started at 2019-09-01T16:40:48.511776
Main binary: /usr/bin/python3
Python system: Linux
The Meson build system
Version: 0.51.2
Source dir: /home/eyeway/cuda-workspace/meson-cuda/src
Build dir: /home/eyeway/cuda-workspace/meson-cuda/build
Build type: native build
Project name: hello-cuda
Project version: undefined
No CXXFLAGS in the environment, not changing global flags.
No LDFLAGS in the environment, not changing global flags.
No CPPFLAGS in the environment, not changing global flags.
Sanity testing C++ compiler: c++
Is cross compiler: False.
Sanity check compiler command line: c++ -pipe -D_FILE_OFFSET_BITS=64 /home/eyeway/cuda-workspace/meson-cuda/build/meson-private/sanitycheckcpp.cc -o /home/eyeway/cuda-workspace/meson-cuda/build/meson-private/sanitycheckcpp.exe
Sanity check compile stdout:

-----
Sanity check compile stderr:

-----
Running test binary command: /home/eyeway/cuda-workspace/meson-cuda/build/meson-private/sanitycheckcpp.exe
No CUFLAGS in the environment, not changing global flags.
No LDFLAGS in the environment, not changing global flags.
Sanity testing Cuda compiler: nvcc
Is cross compiler: None.
Sanity check compiler command line:  nvcc -w -cudart static /home/eyeway/cuda-workspace/meson-cuda/build/meson-private/sanitycheckcuda.cu -o /home/eyeway/cuda-workspace/meson-cuda/build/meson-private/sanitycheckcuda.exe
Sanity check compile stdout: 

-----
Sanity check compile stderr:

-----
Sanity check run command line:  nvcc --run "/home/eyeway/cuda-workspace/meson-cuda/build/meson-private/sanitycheckcuda.exe"
Sanity check run stdout: 
7.2

-----
Sanity check run stderr:

-----
No CXXFLAGS in the environment, not changing global flags.
No LDFLAGS in the environment, not changing global flags.
No CPPFLAGS in the environment, not changing global flags.
Sanity testing C++ compiler: c++
Is cross compiler: False.
Sanity check compiler command line: c++ -pipe -D_FILE_OFFSET_BITS=64 /home/eyeway/cuda-workspace/meson-cuda/build/meson-private/sanitycheckcpp.cc -o /home/eyeway/cuda-workspace/meson-cuda/build/meson-private/sanitycheckcpp.exe
Sanity check compile stdout:

-----
Sanity check compile stderr:

-----
Running test binary command: /home/eyeway/cuda-workspace/meson-cuda/build/meson-private/sanitycheckcpp.exe
C++ compiler for the host machine: c++ (gcc 7.4.0 "c++ (Ubuntu/Linaro 7.4.0-1ubuntu1~18.04.1) 7.4.0")
No CUFLAGS in the environment, not changing global flags.
No LDFLAGS in the environment, not changing global flags.
Sanity testing Cuda compiler: nvcc
Is cross compiler: None.
Sanity check compiler command line:  nvcc -w -cudart static /home/eyeway/cuda-workspace/meson-cuda/build/meson-private/sanitycheckcuda.cu -o /home/eyeway/cuda-workspace/meson-cuda/build/meson-private/sanitycheckcuda.exe
Sanity check compile stdout: 

-----
Sanity check compile stderr:

-----
Sanity check run command line:  nvcc --run "/home/eyeway/cuda-workspace/meson-cuda/build/meson-private/sanitycheckcuda.exe"
Sanity check run stdout: 
7.2

-----
Sanity check run stderr:

-----
Cuda compiler for the host machine: nvcc (nvcc 10.0.166)
Build machine cpu family: aarch64
Build machine cpu: aarch64
meson.build:4: WARNING: Module unstable-cuda has no backwards or forwards compatibility and might not exist in future releases.
Build targets in project: 1
Found ninja-1.8.2 at /usr/bin/ninja
``` 

Here meson correctly identified nvcc as the right compiler. The log in the previous message shows that  from commit 3fbb45016610deba863630b20e95f60d2d49a487 meson doesn't recognize that this is a Cuda build.

**Comment by kpeace** [comment]
I narrowed down the problem to the following lines of code:
environment.py:685
``` python
 def _guess_nix_linker(compiler: typing.List[str], for_machine: MachineChoice,
                          prefix: typing.Union[str, typing.List[str]], *,
                          extra_args: typing.Optional[typing.List[str]] = None) -> 'DynamicLinker':
        """Helper for guessing what linker to use on Unix-Like OSes.
        :prefix: The prefix that the compiler uses to proxy arguments to the
            linker, if required. This can be passed as a string or a list of
            strings. If it is passed as a string then the arguments to be
            proxied to the linker will be concatenated, if it is a list they
            will be appended. This means that if a space is required (such as
            with swift which wants `-Xlinker --version` and *not*
            `-Xlinker=--version`) you must pass as a list.
        :extra_args: Any addtional arguments rquired (such as a source file)
        """
        extra_args = typing.cast(typing.List[str], extra_args or [])
        if isinstance(prefix, str):
            check_args = [prefix + '--version'] + extra_args
        else:
            check_args = prefix + ['--version'] + extra_args
        _, o, e = Popen_safe(compiler + check_args)
        v = search_version(o)
```
Meson is trying to figure out the underlying linker of the compiler. Usually you this by calling the compiler and sending arguments through the compiler to the linker.

nvcc supports sending arguments to the underlying linker BUT will only call the linker after a compilation.

So running:
`nvcc -XLinker=--version`
returns an error. No file to compile specified. (Something along this line...)
Running
`nvcc -Xlinker=--version hello.cu`
will output the the version of the GNU linker (or clang, or whatever).

Adding a file might not solve the problem the problem since nvcc is not a normal compiler.
nvcc
1. compiles only CUDA code
2. is one program that does both the compiling and the linking of the cuda code

So there is no real CUDA linker...

My assumption is that we need a cuda linker object which will not really do anything and a special case for cuda in the above functions which skips the whole "lets figure out the linker".

I would appreciate some input :-)

**Comment by kpeace** [comment]
Fixed with merge #5874. Closing issue.
